### DISCUSSION TITLE
Mingw64(Windows 11)  Failed to built [fixed]

### DISCUSSION THREAD

**Discussion by MinuteMaid20** [bigcat88/pillow_heif#178] [Q&A] [answered]
In Mingw64 (Windows11)  trying to build from source,but something go wrong.
gcc is failed,  maybe missing anything  , source files or libraries.

How can I  do it ?
---What I did for it ---
(1)Uncompressed source files(zip) to temp directory.
Go to libheif/windows/mingw-w64-libheif ---Done
$ pwd
/home/User/temp/pillow_heif-master/libheif/windows/mingw-w64-libheif

(2)Generate Package of libheif  ---Done
$ makepkg-mingw --syncdeps

(3)Installing Mingw64-libheif   Packages  ---Done
$ pacman -U mingw-w64-x86_64-libheif-*-any.pkg.tar.zst

(4) Run setup.py  (pip install) --- [FAILED]
$ pwd
/home/User/temp/pillow_heif-master

$ SETUPTOOLS_USE_DISTUTILS=stdlib MSYS2_PREFIX="C:/msys64/mingw64" pip install .
Processing c:/msys64/home/user/temp/pillow_heif-master
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: pillow>=9.2.0 in c:/msys64/mingw64/lib/python3.11/site-packages (from pillow-heif==0.14.0.dev0) (10.1.0)
Building wheels for collected packages: pillow-heif
  Building wheel for pillow-heif (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for pillow-heif (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [12 lines of output]
      running bdist_wheel
      running build
      running build_py
      running build_ext
      Checking for include file %s in %s ('heif.h', 'C:/msys64/mingw64/include')
      Checking for include file %s in %s ('heif.h', 'C:/msys64/mingw64/include/libheif')
      Found %s in %s ('heif.h', 'C:/msys64/mingw64/include/libheif')
      building '_pillow_heif' extension
      gcc -DNDEBUG -g -fwrapv -O3 -Wall -march=nocona -msahf -mtune=generic -O2 -pipe -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong -O3 -march=nocona -msahf -mtune=generic -O2 -pipe -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong -O3 -IC:/msys64/mingw64/include -IC:/msys64/mingw64/include/python3.11 -IC:/msys64/home/User/temp/pillow_heif-master -c pillow_heif/_pillow_heif.c -o build/temp.mingw_x86_64-3.11/pillow_heif/_pillow_heif.o /d2FH4- /WX
      gcc.exe: fatal error: cannot specify '-o' with '-c', '-S' or '-E' with multiple files
      compilation terminated.
      error: command 'C:\\msys64\\mingw64\\bin/gcc.exe' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pillow-heif
Failed to build pillow-heif
ERROR: Could not build wheels for pillow-heif, which is required to install pyproject.toml-based projects

**Comment by bigcat88** [comment]
I never tried build it with Mingw, only libheif with mingw to get libheif binaries and after that builded "pillow_heif" with `pip install .` from Windows python
I will look today at this situation, never thinked of such...

**Comment by bigcat88** [comment]
With last commit it should succeed to build under Mingw Python, can you verify this?

Ref: https://github.com/bigcat88/pillow_heif/commit/a279fff0f15ef28889483d20963f6662436df3fb

**Comment by MinuteMaid20** [comment, marked as answer]
I'd get it , and succeeded  to build it just now.
Thank you very much.

---Result---
$ pwd
/home/User/temp/1202/pillow_heif-master

User@11-laptop MINGW64 ~/temp/1202/pillow_heif-master
$ ls
CHANGELOG.md          MANIFEST.in  ci      examples  pillow_heif     setup.py
LICENSE.txt           README.md    docker  libheif   pyproject.toml  tests
LICENSES_bundled.txt  benchmarks   docs    pi-heif   setup.cfg

User@11-laptop MINGW64 ~/temp/1202/pillow_heif-master
$ SETUPTOOLS_USE_DISTUTILS=stdlib MSYS2_PREFIX="C:/msys64/mingw64" pip install .
Processing c:/msys64/home/user/temp/1202/pillow_heif-master
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: pillow>=9.2.0 in c:/msys64/mingw64/lib/python3.11/site-packages (from pillow-heif==0.14.0.dev0) (10.1.0)
Building wheels for collected packages: pillow-heif
  Building wheel for pillow-heif (pyproject.toml) ... done
  Created wheel for pillow-heif: filename=pillow_heif-0.14.0.dev0-cp311-cp311-mingw_x86_64.whl size=45038 sha256=2d1a469e02681c6c74473a9812049cdbbdf7815232c2951901c94e41f3b8bc3d
  Stored in directory: c:/users/user/appdata/local/pip/cache/wheels/6f/7f/4d/abf6b50aeb4ea1296add3d39791ef922a30af4879a2d96bdd3
Successfully built pillow-heif
Installing collected packages: pillow-heif
Successfully installed pillow-heif-0.14.0.dev0

$

---Test  --
$ python
Python 3.11.6 (main, Oct  3 2023, 09:12:56)  [GCC 13.2.0 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from PIL import Image, ImageFilter
>>> from pillow_heif import HeifImagePlugin
>>> Image.open("benchmarks/image_large.heic")
<pillow_heif.as_plugin.HeifImageFile image mode=RGB size=8000x6000 at 0x257D812A410>

>>> img = Image.open("benchmarks/image_large.heic")
>>> img.show()
